app/controllers/unattended_controller.rb

```diff
@@ -18,13 +18,27 @@ class UnattendedController < ApplicationController
   # all of our requests should be returned in text/plain
   after_action :set_content_type
 
-  # this actions is called by each operatingsystem post/finish script - it notify us that the OS installation is done.
   def built
     logger.info "#{controller_name}: #{@host.name} is Built!"
+    logger.debug { "Log lines from the OS installer:\n#{extract_build_log}" }
+    # clear possible previous errors
+    @host.build_errors = nil
     update_ip if Setting[:update_ip_from_built_request]
     head(@host.built ? :created : :conflict)
   end
 
+  def failed
+    logger.warn "#{controller_name}: #{@host.name} build Failed!"
+    @host.build_errors = extract_build_log
+    logger.warn { "Log lines from the OS installer:\n#{@host.build_errors}" }
+    head(:created)
+  end
+
+  # last 100 lines from body request
+  def extract_build_log
+    request.body.readlines(200).last(100).join
```

Comment:
Since we are tailing the file, shouldn't there be some message at the beginning that only last 100 lines is captured? I can imagine the confusion, when somebody follows the `Logs from OS installer` link and doesn't see the messages from the beginning, without any indication of the intentional trimming.

Reply:
I found this clunky, rewriting to more efficient code.